src/Microsoft.DotNet.Interactive.CSharp/SignatureHelp/SignatureHelpGenerator.cs

```diff
@@ -0,0 +1,163 @@
+
+            // Walk up until we find a node that we're interested in.
+            while (node != null)
+            {
+                if (node is InvocationExpressionSyntax invocation && invocation.ArgumentList.Span.Contains(position))
+                {
+                    var semanticModel = await document.GetSemanticModelAsync();
+                    return new InvocationContext(semanticModel, position, invocation.Expression, invocation.ArgumentList, invocation.IsInStaticContext());
+                }
+
+                if (node is ObjectCreationExpressionSyntax objectCreation && objectCreation.ArgumentList.Span.Contains(position))
+                {
+                    var semanticModel = await document.GetSemanticModelAsync();
+                    return new InvocationContext(semanticModel, position, objectCreation, objectCreation.ArgumentList, objectCreation.IsInStaticContext());
+                }
+
+                if (node is AttributeSyntax attributeSyntax && attributeSyntax.ArgumentList.Span.Contains(position))
+                {
+                    var semanticModel = await document.GetSemanticModelAsync();
+                    return new InvocationContext(semanticModel, position, attributeSyntax, attributeSyntax.ArgumentList, attributeSyntax.IsInStaticContext());
+                }
+
+                node = node.Parent;
+            }
+
+            return null;
+        }
+
+        private static int InvocationScore(IMethodSymbol symbol, IEnumerable<CodeAnalysis.TypeInfo> types)
+        {
+            var parameters = symbol.Parameters;
+            if (parameters.Count() < types.Count())
+            {
+                return int.MinValue;
+            }
+
+            var score = 0;
+            var invocationEnum = types.GetEnumerator();
+            var definitionEnum = parameters.GetEnumerator();
+            while (invocationEnum.MoveNext() && definitionEnum.MoveNext())
+            {
+                if (invocationEnum.Current.ConvertedType == null)
+                {
+                    // 1 point for having a parameter
+                    score += 1;
+                }
+                else if (SymbolEqualityComparer.Default.Equals(invocationEnum.Current.ConvertedType, definitionEnum.Current.Type))
+                {
+                    // 2 points for having a parameter and being
+                    // the same type
+                    score += 2;
+                }
+            }
+
+            return score;
+        }
+
+        private static SignatureInformation BuildSignature(IMethodSymbol symbol)
+        {
+            var parameters = symbol.Parameters
+                .Select(parameter =>
+                    new ParameterInformation(
+                        label: parameter.Name,
+                        documentation: new FormattedValue("text/markdown", DocumentationConverter.ConvertDocumentation(parameter.GetDocumentationCommentXml(expandIncludes: true)))));
```

Comment:
We talked about deferring this to the `Formatter` so it can be customized more easily and consistently.

Reply:
I think this is a future work item.  The type returned from the Roslyn API is `string` which _might_ be valid XML, and the type eventually returned by F# will be something different.  Eventually we'll need to add `Formatter.Register<SignatureHelpDocumentation>(...)`, but we don't yet have a common type for `SignatureHelpDocumentation` that encompasses everything that might be present in all kinds of sighelp.

Reply:
Got it.